tests/s2_inference/test_automatic_model_ejection.py

```diff
@@ -0,0 +1,214 @@
+        content = "Try to kill the cpu"
+        # These models are tested in encoding test to avoid downloading again.
+        list_of_models = ["fp16/ViT-B/32", "open_clip/convnext_base_w/laion2b_s13b_b82k",
+                          "open_clip/convnext_base_w_320/laion_aesthetic_s13b_b82k_augreg",
+                          "onnx16/open_clip/ViT-B-32/laion400m_e32",
+                          'onnx32/open_clip/ViT-B-32-quickgelu/laion400m_e32',
+                          "all-MiniLM-L6-v1", "all_datasets_v4_MiniLM-L6", "hf/all-MiniLM-L6-v1",
+                          "hf/all_datasets_v4_MiniLM-L6",
+                          "onnx/all-MiniLM-L6-v1", "onnx/all_datasets_v4_MiniLM-L6"]
+
+        for model in list_of_models:
+            _ = vectorise(model_name=model, content=content, device="cpu")
+
+    def test_thread_safe_function_call(self):
+        model_name = "ViT-B/32"
+        validated_model_properties = _validate_model_properties(model_name, None)
+        device = "cpu"
+        model_size = 1
+
+        _validate_model_into_device(model_name, validated_model_properties, device,
+                                    calling_func="_update_available_models")
+        _validate_model_into_device(model_name, validated_model_properties, device, calling_func="unit_test")
+
+        try:
+            _validate_model_into_device(model_name, validated_model_properties, device)
+            raise AssertionError
+        except RuntimeError:
+            pass
+
+        try:
+            _validate_model_into_device(model_name, validated_model_properties, device, calling_func="void")
+            raise AssertionError
+        except RuntimeError:
+            pass
+
+        _check_memory_threshold_for_model(device, model_size, calling_func="_validate_model_into_device")
+        _check_memory_threshold_for_model(device, model_size, calling_func="unit_test")
+
+        try:
+            _check_memory_threshold_for_model(device, model_size)
+            raise AssertionError
+        except RuntimeError:
+            pass
+
+        try:
+            _check_memory_threshold_for_model(device, model_size, calling_func="void")
+            raise AssertionError
+        except RuntimeError:
+            pass
+
+    def test_concurrent_vectorise_call_no_cache(self):
+        # To test error is thrown if multiple threads want to load the model
+        clear_loaded_models()
+        test_content = "this is a test"
+        test_model = "ViT-B/32"
+
+        threads = []
+        q = queue.Queue()
+        t = threading.Thread(target=normal_vectorise_call, args=(test_model, test_content, q))
+        threads.append(t)
+        t.start()
+        for i in range(3):
+            t = threading.Thread(target=racing_vectorise_call, args=(test_model, test_content, q))
+            threads.append(t)
+            t.start()
+
+        for t in threads:
+            t.join()
+
+        assert q.empty()
```

Comment:
my preference is to put model_load_rejection errors into the shared data structure (here a queue), and then assert that each error is the correct type. That way we can assert about the expected number of errors and that they re the correct error type (for example, perhaps q remains empty because the test doesn't work)